vsintegration/src/FSharp.Editor/Completion/CompletionProvider.fs

```diff
@@ -219,24 +219,43 @@ type internal FSharpCompletionProvider
-        asyncMaybe {
-            use _logBlock = Logger.LogBlockMessage context.Document.Name LogEditorFunctionId.Completion_ProvideCompletionsAsync
+        async {
+            Logger.LogBlockMessageStart context.Document.Name LogEditorFunctionId.Completion_ProvideCompletionsAsync
 
             let document = context.Document
-            let! sourceText = context.Document.GetTextAsync(context.CancellationToken)
+            let! sourceText = context.Document.GetTextAsync(context.CancellationToken) |> Async.AwaitTask
+
+            match sourceText with
+            | null -> Logger.LogBlockMessageStop (context.Document.Name + "-Failed_SourceTextNotFound") LogEditorFunctionId.Completion_ProvideCompletionsAsync
+            | _ ->
+
             let defines = projectInfoManager.GetCompilationDefinesForEditingDocument(document)
-            do! Option.guard (CompletionUtils.shouldProvideCompletion(document.Id, document.FilePath, defines, sourceText, context.Position))
-            let! _parsingOptions, projectOptions = projectInfoManager.TryGetOptionsForEditingDocumentOrProject(document)
-            let! textVersion = context.Document.GetTextVersionAsync(context.CancellationToken)
+
+            if not (CompletionUtils.shouldProvideCompletion(document.Id, document.FilePath, defines, sourceText, context.Position)) then
+                Logger.LogBlockMessageStop (context.Document.Name + "-Failed_ShouldProvideCompletionReturnedFalse") LogEditorFunctionId.Completion_ProvideCompletionsAsync
+            else
+
+            match projectInfoManager.TryGetOptionsForEditingDocumentOrProject(document) with
+            | None -> Logger.LogBlockMessageStop (context.Document.Name + "-Failed_OptionsNotFound") LogEditorFunctionId.Completion_ProvideCompletionsAsync
+            | Some(_, projectOptions) ->
+
+            let! textVersion = context.Document.GetTextVersionAsync(context.CancellationToken) |> Async.AwaitTask
             let getAllSymbols(fileCheckResults: FSharpCheckFileResults) =
                 if Settings.IntelliSense.ShowAllSymbols
                 then assemblyContentProvider.GetAllEntitiesInProjectAndReferencedAssemblies(fileCheckResults)
                 else []
-            let! results = 
-                FSharpCompletionProvider.ProvideCompletionsAsyncAux(checker, sourceText, context.Position, projectOptions, 
-                                                                    document.FilePath, textVersion.GetHashCode(), getAllSymbols)
+
+            let! resultsOpt = FSharpCompletionProvider.ProvideCompletionsAsyncAux(checker, sourceText, context.Position, projectOptions, 
+                                                                                  document.FilePath, textVersion.GetHashCode(), getAllSymbols)
+
+            match resultsOpt with
+            | None -> Logger.LogBlockMessageStop "-Failed_ProvideCompeltionsAsyncAuxFailed" LogEditorFunctionId.Completion_ProvideCompletionsAsync
```

Comment:
Why we're adding all this logging code?  I mean, I know it must be important.  But it really messes with the readability of the code (I used to find this code very declarative and even used it as an example of great F# code in a recent talk.  Now it's really hard to read....)

Can't we find some more declarative way of adding this?  e.g. make an `asyncMaybeLogging` computation expression?

Reply:
Typo in spelling of "Compeltion"

Reply:
It wasn't super important as I never get logs for them, I put them in there because I wanted to have detailed logs describing why something failed. Using the `use` pattern for logging in which things will fail or short circuit execution, I never know why they would based on the logs; I would just see that it stopped quicker than usual.

However, most of these never get hit when I look at the logs. I assumed that it would be ok to keep these in here just in case.

I'll see if I can find a better way here.